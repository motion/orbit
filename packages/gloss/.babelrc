{
  "env": {
    "development": {
      "presets": [
        [
          "@o/babel-preset-motion",
          {
            "glossDisplayNameConfig": {
              "matchImports": ["../gloss", "."]
            }
          }
        ]
      ]
    }
  }
}
