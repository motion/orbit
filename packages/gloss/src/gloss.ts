import css from '@mcro/css'
import { Options } from './types'
import { createViewFactory } from './createViewFactory'
import { GLOSS_IGNORE_COMPONENT_SYMBOL } from './symbols'

export { GLOSS_IGNORE_COMPONENT_SYMBOL } from './symbols'
export { ThemeObject } from './types'
export { alphaColor } from './helpers/alphaColor'
export { propsToStyles } from './helpers/propsToStyles'
export { propsToThemeStyles } from './helpers/propsToThemeStyles'
export { Theme } from './theme/Theme'
export { ThemeMaker } from './theme/ThemeMaker'
export { ThemeProvide } from './theme/ThemeProvide'
export { ThemeContext } from './theme/ThemeContext'
export { attachTheme } from './theme/attachTheme'
export {
  CSSPropertySet,
  cssNameMap,
  psuedoKeys,
  validCSSAttr,
  Color,
} from '@mcro/css'

export const color = require('@mcro/color').default

export const colorToString = obj => {
  const { model, color, valpha } = obj
  const hasAlpha = typeof valpha === 'number' && valpha !== 1
  if (model === 'rgb') {
    const inner = `${color[0]}, ${color[1]}, ${color[2]}`
    if (hasAlpha) {
      return `rgba(${inner}, ${valpha})`
    }
    return `rgb(${inner})`
  }
  if (model === 'hsl') {
    const inner = `${color[0]}, ${Math.round(color[1])}%, ${Math.round(
      color[2],
    )}%`
    if (hasAlpha) {
      return `hsla(${inner}, ${valpha})`
    }
    return `hsl(${inner})`
  }
  return obj.toString()
}

const isGlossFirstArg = a =>
  typeof a === 'undefined' || typeof a === 'string' || typeof a === 'object'

export const isGlossArguments = (a, b) => {
  if (b && b[GLOSS_IGNORE_COMPONENT_SYMBOL]) {
    return false
  }
  if (typeof a === 'function' && typeof b === 'object') {
    return true
  }
  return isGlossFirstArg(a)
}

const DEFAULT_OPTS = {
  glossProp: 'css',
  isColor: color => color && !!color.rgb,
  toColor: colorToString,
}

export default function createGloss(options: Options = DEFAULT_OPTS) {
  const createStyles = css(options)
  const createView = createViewFactory(createStyles)
  return {
    createStyles,
    createView,
  }
}
