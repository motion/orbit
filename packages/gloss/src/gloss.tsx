import { CSSPropertySet, CSSPropertySetLoose, cssString, styleToClassName, ThemeObject, validCSSAttr } from '@o/css'
import { createElement, forwardRef, isValidElement, useEffect, useRef } from 'react'

import { Config } from './config'
import { useTheme } from './helpers/useTheme'
import { validProp } from './helpers/validProp'
import { GarbageCollector, StyleTracker } from './stylesheet/gc'
import { StyleSheet } from './stylesheet/sheet'
import { ThemeSelect } from './theme/Theme'

export const baseIgnoreAttrs = {
  ...validCSSAttr,
  width: false,
  height: false,
  size: false,
  src: false,
}

export type BaseRules = {
  [key: string]: string | number
}

export type GlossProps<Props> = Props & {
  className?: string
  tagName?: string
  children?: React.ReactNode
  ref?: any
  style?: any
  alt?: string
  subTheme?: ThemeSelect
}

export type ThemeFn<Props = any> = (
  props: GlossProps<Props>,
  theme: ThemeObject,
  previous?: CSSPropertySetLoose | null,
) => CSSPropertySetLoose | undefined | null

export type GlossConfig<Props> = {
  displayName?: string
  ignoreAttrs?: { [key: string]: boolean }
  defaultProps?: Partial<Props>
  modifyProps?: (curProps: Props, nextProps: any) => any
  getElement?: (props: Props) => any
  isDOMElement?: boolean
}

type GlossInternalConfig = {
  id: string
  displayName: string
  targetElement: any
  styles: any
  conditionalStyles: Object
  config: GlossConfig<any> | null
}

type GlossInternals<Props> = {
  parent: any
  themeFns: ThemeFn<Props>[] | null
  getConfig: () => GlossInternalConfig
}

/**
 * Note: ThemeProps is optional, for the user to define that they are
 * filtering out props before they get to `theme`, which is helpful for keeping
 * their types simple. It's still relatively low level. See our @o/ui/View.tsx
 */
export interface GlossView<RawProps, ThemeProps = RawProps, Props = GlossProps<RawProps>> {
  // copied from FunctionComponent
  (props: Props, context?: any): React.ReactElement<any> | null
  propTypes?: React.ValidationMap<Props>
  contextTypes?: React.ValidationMap<any>
  defaultProps?: Partial<Props>
  displayName?: string
  // extra:
  ignoreAttrs?: { [key: string]: boolean }
  theme: (...themeFns: ThemeFn<ThemeProps>[]) => GlossView<RawProps>
  withConfig: (config: GlossConfig<Props>) => GlossView<RawProps>
  internal: GlossInternals<Props>
}

const GLOSS_SIMPLE_COMPONENT_SYMBOL = '__GLOSS_SIMPLE_COMPONENT__'
export const tracker: StyleTracker = new Map()
export const sheet = new StyleSheet(true)
const rulesToClass = new WeakMap()
const gc = new GarbageCollector(sheet, tracker, rulesToClass)
const whiteSpaceRegex = /[\s]+/g

let idCounter = 1
const viewId = () => idCounter++ % Number.MAX_SAFE_INTEGER

export function gloss<Props = any, ThemeProps = Props>(
  a?: CSSPropertySet | GlossView<Props, ThemeProps> | ((props: Props) => any) | string,
  b?: CSSPropertySet,
): GlossView<GlossProps<Props>, ThemeProps> {
  let target: any = a || 'div'
  let rawStyles = b
  let ignoreAttrs: Object
  const hasGlossyChild = !!target[GLOSS_SIMPLE_COMPONENT_SYMBOL]
  const targetConfig: GlossInternalConfig | null = hasGlossyChild
    ? target.internal.getConfig()
    : null

  setTimeout(() => {
    if (!ignoreAttrs) {
      ignoreAttrs =
        ThemedView.ignoreAttrs || (hasGlossyChild && target.ignoreAttrs) || baseIgnoreAttrs
    }
  }, 0)

  // shorthand: gloss({ ... })
  if (
    typeof a !== 'string' &&
    target.constructor &&
    target.constructor.name === 'Object' &&
    !b &&
    !hasGlossyChild &&
    !isValidElement(target) &&
    // if you pass an element wrapped in forwardRef() it needs to be filtered somehow..
    !target['$$typeof']
  ) {
    target = 'div'
    rawStyles = a
  }

  const targetElement = !!targetConfig ? targetConfig.targetElement : target
  const targetElementName = typeof targetElement === 'string' ? targetElement : ''
  const id = `${viewId()}`
  const Styles = getAllStyles(id, targetConfig, rawStyles || null)

  const conditionalStyles = Styles.conditionalStyles
  let themeFn: ThemeFn | null = null
  let staticClasses: string[] | null = null

  // this elements helpers
  let ogConfig: GlossConfig<Props> | null = null
  let config: GlossConfig<Props> | null = null

  let hasCompiled = false

  /**
   *
   *
   * ðŸ’… The component for each gloss view.
   *
   *
   */
  function GlossView(props: GlossProps<Props>, ref: any) {
    // compile on first run to avoid extra work
    if (!hasCompiled) {
      hasCompiled = true
      themeFn = compileTheme(ThemedView)
      staticClasses = addStyles(Styles.styles, ThemedView.displayName, targetElementName)
      config = getCompiledConfig(ThemedView, ogConfig)
    }

    const theme = useTheme()
    const dynClasses = useRef<Set<string> | null>(null)

    // unmount
    useEffect(() => {
      return () => {
        const x = dynClasses.current
        if (x && x.size > 0) {
          x.forEach(deregisterClassName)
        }
      }
    }, [])

    const dynClassNames = addDynamicStyles(
      id,
      ThemedView.displayName,
      conditionalStyles,
      dynClasses.current,
      props,
      themeFn,
      theme,
      targetElementName,
    )

    const classNames = new Set<string>(
      staticClasses
        ? dynClassNames
          ? [...staticClasses, ...dynClassNames]
          : staticClasses
        : dynClassNames,
    )

    dynClasses.current = dynClassNames

    // if this is a plain view we can use tagName, otherwise just pass it down
    let element = typeof targetElement === 'string' ? props.tagName || targetElement : targetElement

    // helper for element
    const getEl = config && config.getElement
    if (getEl) {
      element = getEl(props)
    }

    const isDOMElement = typeof element === 'string' || (config && config.isDOMElement)

    // set up final props with filtering for various attributes
    let finalProps: any = {}

    if (ref) {
      finalProps.ref = ref
    }

    if (isDOMElement) {
      for (const key in props) {
        if (ignoreAttrs && ignoreAttrs[key]) continue
        if (conditionalStyles && conditionalStyles[key]) continue
        // TODO: need to figure out this use case: when a valid prop attr, but invalid val
        if (key === 'size' && typeof props[key] !== 'string') continue
        if (!validProp(key)) continue
        finalProps[key] = finalProps[key] || props[key]
      }
    } else {
      for (const key in props) {
        if (conditionalStyles && conditionalStyles[key]) continue
        finalProps[key] = finalProps[key] || props[key]
      }
    }

    // we control these props
    finalProps.className = [...classNames].join(' ')
    finalProps['data-is'] = finalProps['data-is'] || ThemedView.displayName

    // hook: setting your own props
    const modifyProps = config && config.modifyProps
    if (modifyProps) {
      modifyProps(props, finalProps)
    }

    return createElement(element, finalProps, props.children)
  }

  const internal: GlossInternals<Props> = {
    themeFns: null,
    parent: hasGlossyChild ? target : null,
    getConfig: () => ({
      config: ogConfig,
      id,
      displayName: ThemedView.displayName || '',
      targetElement,
      styles: { ...Styles.styles },
      conditionalStyles: { ...Styles.conditionalStyles },
    }),
  }

  let ThemedView = createGlossView<Props>(GlossView, internal)

  // inherit default props
  if (hasGlossyChild) {
    ThemedView.defaultProps = target.defaultProps
  }

  ThemedView.withConfig = opts => {
    ogConfig = ogConfig || {}
    Object.assign(ogConfig, opts)

    // re-create it so it picks up displayName
    if (opts.displayName) {
      // this one is picked up by Profiling
      GlossView['displayName'] = opts.displayName
      ThemedView = createGlossView<Props>(GlossView, internal)
      // this one is picked up for use in classNames
      ThemedView['displayName'] = opts.displayName
    }

    if (opts.ignoreAttrs) {
      // TODO could have option to override or merge
      ignoreAttrs = {
        ...baseIgnoreAttrs,
        ...opts.ignoreAttrs,
      }
    }

    if (opts.defaultProps) {
      ThemedView.defaultProps = opts.defaultProps
    }

    return ThemedView
  }

  // TODO this any type is a regression from adding ThemeProps
  return ThemedView as any
}

function createGlossView<Props>(GlossView: any, config) {
  const res: GlossView<Props> = forwardRef<HTMLDivElement, GlossProps<Props>>(GlossView) as any
  res.internal = config
  res[GLOSS_SIMPLE_COMPONENT_SYMBOL] = true
  res.theme = (...themeFns) => {
    config.themeFns = themeFns
    return res
  }
  return res
}

// keeps priority of hover/active/focus as expected
const psuedoScore = (x: string) => {
  const hasFocus = x.indexOf('&:focus') > -1 ? 1 : 0
  const hasHover = x.indexOf('&:hover') > -1 ? 2 : 0
  const hasActive = x.indexOf('&:active') > -1 ? 3 : 0
  return hasActive + hasHover + hasFocus
}

const pseudoSort = (a: string, b: string) => (psuedoScore(a) > psuedoScore(b) ? 1 : -1)

// takes a style object, adds it to stylesheet, returns classnames
function addStyles(
  styles: any,
  displayName?: string,
  tagName?: string,
  prevClassNames?: Set<string> | null,
  moreSpecific?: boolean,
) {
  const keys = Object.keys(styles).sort(pseudoSort)
  let classNames: string[] | null = null
  for (const key of keys) {
    const rules = styles[key]
    // they may return falsy, conditional '&:hover': active ? hoverStyle : null
    if (!rules) continue

    // add the stylesheets and classNames
    // TODO this could do a simple "diff" so that fast-changing styles only change the "changing" props
    // it would likely help things like when you animate based on mousemove, may be slower in default case
    const className = addRules(displayName, rules, key, tagName, moreSpecific)
    classNames = classNames || []
    classNames.push(className)

    // if this is the first mount render or we didn't previously have this class then add it as new
    if (!prevClassNames || !prevClassNames.has(className)) {
      gc.registerClassUse(className)
    }
  }
  return classNames
}

function mergePropStyles(baseId: string, styles: Object, propStyles: Object, props: Object) {
  for (const key in propStyles) {
    if (props[key] !== true) continue
    for (const sKey in propStyles[key]) {
      const ns = sKey === 'base' ? baseId : sKey
      styles[ns] = styles[ns] || {}
      mergeStyles(ns, styles, propStyles[key][sKey], true)
    }
  }
}

const SPECIFIC_PREFIX = 's'

function deregisterClassName(name: string) {
  const nonSpecificClassName = name[0] === SPECIFIC_PREFIX ? name.slice(1) : name
  gc.deregisterClassUse(nonSpecificClassName)
}

const isNumericString = (x: string | number) => +x == x

function addDynamicStyles(
  id: string,
  displayName: string = 'g',
  conditionalStyles: Object | undefined,
  prevClassNames: Set<string> | null,
  props: CSSPropertySet,
  themeFn?: ThemeFn | null,
  theme?: ThemeObject,
  tagName?: string,
) {
  const dynStyles = {}
  let classNames = new Set<string>()

  // applies styles most important to least important
  // that saves us some processing time (no need to set multiple times)
  // note that means the topmost `mergeStyles` will apply as most important

  // if passed any classes from a parent gloss view, merge them, ignore classname and track
  if (props.className) {
    const propClassNames = props.className.trim().split(whiteSpaceRegex)
    // note this reverse: this is a bit odd
    // right now we have conditionalStyles applied as their own className (so base: .1, conditional: .2)
    // then we pass className="1 2" if we have a parent that the conditional style === true
    // what we probably want is to merge them all into their own single className
    // until then, we need to preserve the important order, so we reverse to make sure conditional applies first
    propClassNames.reverse()
    for (const className of propClassNames) {
      const cn = className[0] === SPECIFIC_PREFIX ? className.slice(1) : className
      const info = tracker.get(cn)
      if (info) {
        // curId is looking if info.namespace was &:hover (sub-select) or "123" (base) and then applying
        // otherwise it would apply hover styles to the base styles here
        const curId = isNumericString(info.namespace) ? id : info.namespace
        dynStyles[curId] = dynStyles[curId] || {}
        mergeStyles(curId, dynStyles, info.rules)
      } else if (className) {
        classNames.add(className)
      }
    }
  }

  if (conditionalStyles) {
    mergePropStyles(id, dynStyles, conditionalStyles, props)
  }

  if (theme && themeFn) {
    const next = Config.preProcessTheme ? Config.preProcessTheme(props, theme) : theme
    dynStyles[id] = dynStyles[id] || {}
    const themeStyles = themeFn(props, next)
    const themePropStyles = mergeStyles(id, dynStyles, themeStyles, true)
    if (themePropStyles) {
      mergePropStyles(id, dynStyles, themePropStyles, props)
    }
  }

  // add dyn styles
  const dynClassNames = addStyles(dynStyles, displayName, tagName, prevClassNames, true)
  if (dynClassNames) {
    for (const cn of dynClassNames) {
      classNames.add(cn)
    }
  }

  // check what classNames have been removed if this is a secondary render
  if (prevClassNames) {
    for (const className of prevClassNames) {
      // if this previous class isn't in the current classes then deregister it
      if (!classNames.has(className)) {
        deregisterClassName(className)
      }
    }
  }

  return classNames
}

const isSubStyle = (x: string) => x[0] === '&' || x[0] === '@'

//
// this... THIS...
//  ... this is a tricky function
//  because its used on initial mount AND during renders
//  which is actually useful, yes, because you want the logic the same
//  BUT its also used nested! See themeFn => mergePropStyles
//  likely can be refactored, but just need to study it a bit before you do
//
function mergeStyles(
  id: string,
  baseStyles: Object,
  nextStyles?: CSSPropertySet | null,
  overwrite?: boolean,
): Object | undefined {
  let propStyles
  for (const key in nextStyles) {
    // dont overwrite as we go down
    if (overwrite !== true && baseStyles[id][key] !== undefined) {
      continue
    }
    if (validCSSAttr[key]) {
      // valid regular attr
      baseStyles[id][key] = nextStyles[key]
    } else if (isSubStyle(key)) {
      for (const sKey in nextStyles[key]) {
        if (overwrite === true || !baseStyles[key] || baseStyles[key][sKey] === undefined) {
          baseStyles[key] = baseStyles[key] || {}
          baseStyles[key][sKey] = nextStyles[key][sKey]
        }
      }
    } else {
      let subStyles = nextStyles[key]
      // catch invalid/falsy
      if (!subStyles || typeof subStyles !== 'object') {
        continue
      }
      // propStyles
      //   definition: gloss({ isTall: { height: '100%' } })
      //   usage: <Component isTall />

      propStyles = propStyles || {}
      propStyles[key] = {}

      // they can nest (media queries/psuedoes), split it out, eg:
      //  gloss({
      //    isTall: {
      //      '&:before': {}
      //    }
      //  })
      for (const sKey in subStyles) {
        // key = isTall
        // sKey = &:before
        if (isSubStyle(sKey)) {
          // keep all sub-styles on their key
          propStyles[key] = propStyles[key] || {}
          propStyles[key][sKey] = subStyles[sKey]
        } else {
          // we put base styles here, see 'base' check above
          propStyles[key].base = propStyles[key].base || {}
          propStyles[key].base[sKey] = subStyles[sKey]
        }
      }
    }
  }
  return propStyles
}

// happens once at initial gloss() call, so not as perf intense
// get all parent styles and merge them into a big object
function getAllStyles(
  baseId: string,
  config: GlossInternalConfig | null,
  rawStyles: CSSPropertySet | null,
) {
  const styles = {
    [baseId]: {},
  }
  let conditionalStyles = mergeStyles(baseId, styles, rawStyles)
  // merge parent styles
  if (config) {
    const parentPropStyles = config.conditionalStyles
    if (parentPropStyles) {
      for (const key in parentPropStyles) {
        conditionalStyles = conditionalStyles || {}
        conditionalStyles[key] = conditionalStyles[key] || {}
        conditionalStyles[key] = {
          ...parentPropStyles[key],
          ...conditionalStyles[key],
        }
      }
    }
    const parentStyles = config.styles
    const parentId = config.id
    const moveToMyId = parentStyles[parentId]
    delete parentStyles[parentId]
    parentStyles[baseId] = moveToMyId
    for (const key in parentStyles) {
      styles[key] = {
        ...parentStyles[key],
        ...styles[key],
      }
    }
  }
  return {
    styles,
    conditionalStyles,
  }
}

/**
 * We need to compile a few things to get the config right:
 *   1. get all the parents modifyProps until:
 *   2. encounter a parent with getElement (and use that isDOMElement)
 *   3. stop there, don't keep going higher
 */
function getCompiledConfig(
  viewOG: GlossView<any>,
  config: GlossConfig<any> | null,
): GlossConfig<any> {
  const compiledConf: GlossConfig<any> = { ...config }
  let cur = viewOG
  while (cur) {
    const curConf = cur.internal.getConfig().config
    if (curConf) {
      if (curConf.modifyProps) {
        // merge the modifyProps
        const og = compiledConf.modifyProps
        compiledConf.modifyProps = og
          ? (a, b) => {
              og(a, b)
              curConf.modifyProps!(a, b)
            }
          : curConf.modifyProps
      }
      // find the first getElement and break here
      if (curConf.getElement) {
        compiledConf.getElement = curConf.getElement
        compiledConf.isDOMElement = curConf.isDOMElement
        break
      }
    }
    cur = cur.internal.parent
  }
  return compiledConf
}

// compile theme from parents
function compileTheme(viewOG: GlossView<any>) {
  let cur = viewOG
  let all: ThemeFn[][] = []

  // get themes in order from most important (current) to least important (grandparent)
  while (cur) {
    const conf = cur.internal
    if (conf.themeFns) {
      all.push(conf.themeFns)
    }
    cur = conf.parent
  }

  // reverse, then flatten, so its a flat list of themes from least to most important
  // makes it easier to apply them in order
  all.reverse()
  const themes = all.flat().filter(Boolean)

  if (!themes.length) {
    return null
  }

  return (props: Object, theme: ThemeObject) => {
    let styles: CSSPropertySetLoose | null = null
    for (const themeFn of themes) {
      const next = themeFn(props, theme, styles)
      if (next) {
        styles = styles || {}
        Object.assign(styles, next)
      }
    }
    return styles
  }
}

// adds rules to stylesheet and returns classname
function addRules(
  displayName = '_',
  rules: BaseRules,
  namespace: string,
  tagName?: string,
  moreSpecific?: boolean,
) {
  // if these rules have been cached to a className then retrieve it
  const cachedClass = rulesToClass.get(rules)
  if (cachedClass) {
    return cachedClass
  }

  const style = cssString(rules)

  // build the class name with the display name of the styled component and a unique id based on the css and namespace
  const className = styleToClassName(style)

  // this is the first time we've found this className
  if (!tracker.has(className)) {
    // build up the correct selector, explode on commas to allow multiple selectors
    const selector = getSelector(className, namespace, tagName)
    // insert the new style text
    tracker.set(className, {
      displayName,
      namespace,
      rules,
      selector,
      style,
      className,
    })

    if (namespace[0] === '@') {
      sheet.insert(namespace, `${namespace} {\n${selector} {\n${style}\n}\n}`)
    } else {
      sheet.insert(className, `${selector} {\n${style}\n}`)
    }

    rulesToClass.set(rules, className)
  }

  return moreSpecific ? `${SPECIFIC_PREFIX}${className}` : className
}
// has to return a .s-id and .id selector for use in parents passing down styles
function getSelector(className: string, namespace: string, tagName: string = '') {
  if (namespace[0] === '@') {
    return `${tagName}.${className}, body ${tagName}.${SPECIFIC_PREFIX}${className}`
  }
  if (namespace.indexOf('&') !== -1) {
    // namespace === '&:hover, &:focus, & > div'
    const namespacedSelectors = namespace
      .split(',')
      .flatMap(part => {
        const selector = part.replace('&', className).trim()
        return [`body .${SPECIFIC_PREFIX}${selector}`, `.${selector}`]
      })
      .join(',')
    return namespacedSelectors
  }
  return `body .${SPECIFIC_PREFIX}${className}, .${className}`
}
