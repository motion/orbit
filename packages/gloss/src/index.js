// @flow
import fancyElFactory from './fancyElement'
import { StyleSheet } from './stylesheet'
import { pickBy } from 'lodash'
import { applyNiceStyles, flattenThemes, isFunc } from './helpers'

// exports
export { colorToString, objectToColor, expandCSSArray } from 'motion-css'
export type { Transform, Color } from 'motion-css'
// export components
export ThemeProvide from './components/themeProvide'
export Theme from './components/theme'

function getStyles({ name, style }, theme: ?Object) {
  const styles = { ...style, ...flattenThemes(theme) }
  const dynamicStyles = pickBy(styles, isFunc)
  const staticStyles = pickBy(styles, x => !isFunc(x))
  const niceStatics = applyNiceStyles(staticStyles, `${name}:`)
  const statics = StyleSheet.create(niceStatics)
  // attach key to status objects so we can use later in fancyElement
  for (const key of Object.keys(statics)) {
    statics[key].key = key
  }
  return {
    statics,
    dynamics: dynamicStyles,
    theme,
  }
}

const DEFAULT_OPTS = {
  themeKey: 'theme',
}

export default function glossFactory(opts: Object = DEFAULT_OPTS): Function {
  let baseStyles
  if (opts.baseStyles) {
    baseStyles = getStyles(
      { name: 'Gloss Parent Styles', style: opts.baseStyles },
      null
    )
  }

  const fancyEl = (styles, theme) =>
    fancyElFactory(theme, baseStyles, styles, opts)

  function gloss(ChildOrName: Function | string, style: ?Object) {
    // shorthand
    if (typeof ChildOrName === 'string') {
      const name = ChildOrName
      const createEl = fancyEl(getStyles({ name, style: { [name]: style } }))
      return ({ getRef, ...props }) => createEl(name, { ref: getRef, ...props })
    }

    // using as decorator for class
    const Child = ChildOrName

    // shim this.fancyElement
    if (Child.prototype) {
      Child.prototype.glossElement = fancyEl(
        getStyles(Child, opts.dontTheme ? null : Child.theme),
        Child.theme
      )
    }

    return Child
  }

  // allow access directly to Gloss.createElement
  gloss.createElement = fancyEl(getStyles({ name: 'root', style: {} }))

  return gloss
}
