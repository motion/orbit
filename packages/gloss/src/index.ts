export {
  configureCSS,
  css,
  CSSPropertySet,
  CSSPropertySetStrict,
  psuedoKeys,
  ThemeObject,
  validCSSAttr,
} from '@o/css'
export * from './blocks/Absolute'
export * from './blocks/Base'
// blocks
export * from './blocks/Block'
export * from './blocks/Col'
export * from './blocks/Contents'
export * from './blocks/FullScreen'
export * from './blocks/Grid'
export * from './blocks/Inherit'
export * from './blocks/Inline'
export * from './blocks/InlineBlock'
export * from './blocks/InlineFlex'
export * from './blocks/Row'
// configureGloss
export { configureGloss } from './config'
export { gloss, GlossProps, GlossView, ThemeFn } from './gloss'
export { colorToString } from './helpers/helpers'
export { preProcessTheme } from './helpers/preProcessTheme'
export { selectThemeSubset } from './helpers/selectThemeSubset'
export { useTheme, useThemeContext } from './helpers/useTheme'
export { attachTheme } from './theme/attachTheme'
export { Theme, ThemeSelect } from './theme/Theme'
export { ThemeContext, ThemeContextType } from './theme/ThemeContext'
export { ThemeProvide } from './theme/ThemeProvide'
export * from './themes'
export { propStyleTheme as propsToStyles } from './themes/propStyleTheme'
export { SimpleStyleObject } from './types'
