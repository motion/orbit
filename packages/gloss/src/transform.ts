import * as babel from '@babel/core'
import { SourceMapGenerator } from 'source-map'

import { loadOptions, PluginOptions } from './babel/loadOptions'

export type Replacement = {
  original: { start: Location; end: Location }
  length: number
}

type Location = {
  line: number
  column: number
}

type Result = {
  code: string
  sourceMap?: Object
  cssText?: string
  cssSourceMapText?: string
  dependencies?: string[]
  rules?: {
    [className: string]: {
      cssText: string
      displayName: string
      start?: Location
    }
  }
  replacements?: Replacement[]
}

type Options = {
  filename: string
  preprocessor?: Preprocessor
  outputFilename?: string
  inputSourceMap?: Object
  pluginOptions?: PluginOptions
}

export type Preprocessor =
  | 'none'
  | 'stylis'
  | ((selector: string, cssText: string) => string)
  | void

export async function transform(code: string, options: Options): Promise<Result> {
  const pluginOptions = await loadOptions(options.pluginOptions)

  // Parse the code first so babel uses user's babel config for parsing
  // We don't want to use user's config when transforming the code
  const ast = babel.parseSync(code, {
    ...(pluginOptions ? pluginOptions.babelOptions : null),
    filename: options.filename,
    caller: { name: 'gloss' },
  })

  if (!ast) {
    throw new Error('No AST')
  }

  const res = babel.transformFromAstSync(ast, code, {
    filename: options.filename,
    presets: [[require.resolve('./babel'), pluginOptions]],
    babelrc: false,
    configFile: false,
    sourceMaps: true,
    sourceFileName: options.filename,
    inputSourceMap: options.inputSourceMap,
  })

  if (!res) {
    throw new Error('No babel res')
  }

  const { metadata, code: transformedCode, map } = res

  console.log('metadata', metadata)

  const out = metadata && metadata['gloss']

  if (!out) {
    return {
      code,
      sourceMap: options.inputSourceMap,
    }
  }

  if (!map) {
    throw new Error('No map returned')
  }

  const { rules, replacements, dependencies } = out
  const mappings: Object[] = []

  let cssText = ''

  Object.keys(rules).forEach((selector, index) => {
    mappings.push({
      generated: {
        line: index + 1,
        column: 0,
      },
      original: rules[selector].start,
      name: selector,
    })

    // Run each rule through stylis to support nesting
    cssText += `${rules[selector].cssText}\n`
  })

  return {
    code: transformedCode || '',
    cssText,
    rules,
    replacements,
    dependencies,
    sourceMap: map,
    get cssSourceMapText() {
      if (mappings && mappings.length) {
        const generator = new SourceMapGenerator({
          file: options.filename.replace(/\.js$/, '.css'),
        })

        mappings.forEach(mapping =>
          generator.addMapping({ ...mapping, source: options.filename } as any),
        )

        generator.setSourceContent(options.filename, code)

        return generator.toString()
      }

      return ''
    },
  }
}
