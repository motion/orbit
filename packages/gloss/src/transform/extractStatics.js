// @flow
export default function({ types: t }: { types: Object }) {
  return {
    visitor: {
      Class(path: Object, state: Object) {
        const props = []
        const body = path.get('body')

        for (const path of body) {
          if (path.isClassProperty()) {
            console.log(path)
            props.push(path)
          }
        }
      },
    },
  }
}
