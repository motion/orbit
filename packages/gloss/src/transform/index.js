// @flow
import helper from 'babel-helper-builder-react-jsx'

export default function({ types: t }: { types: Object }) {
  // convert React.createElement() => this.glossElement()

  // todo this is horrible and lame
  const isIf = attribute =>
    attribute.type === 'JSXAttribute' && attribute.name.name === 'if'
  const isntIf = x => !isIf(x)
  const jsxIfPlugin = path => {
    const { node } = path
    const attributes = node.openingElement.attributes
    if (!attributes) return
    const ifAttribute = attributes.filter(isIf)[0]
    if (ifAttribute) {
      const opening = t.JSXOpeningElement(
        node.openingElement.name,
        attributes.filter(isntIf)
      )
      const tag = t.JSXElement(opening, node.closingElement, node.children)
      const conditional = t.conditionalExpression(
        ifAttribute.value.expression,
        tag,
        t.nullLiteral()
      )
      path.replaceWith(conditional)
    }
  }

  const classBodyVisitor = {
    ClassMethod(path: Object, state: Object) {
      const GLOSS_ID = path.scope.generateUidIdentifier('gloss')
      let hasJSX = false

      const { JSXNamespacedName, JSXElement } = helper({
        post(state) {
          // need path to determine if variable or tag
          const stupidIsTag =
            state.tagName && state.tagName[0].toLowerCase() === state.tagName[0]

          state.call = t.callExpression(GLOSS_ID, [
            stupidIsTag ? t.stringLiteral(state.tagName) : state.tagExpr,
            ...state.args,
          ])
        },
      })

      path.traverse(
        {
          JSXNamespacedName,
          JSXElement: {
            enter(...args) {
              if (state.opts.jsxIf) {
                jsxIfPlugin(...args)
              }
              hasJSX = true
            },
            exit: JSXElement.exit,
          },
        },
        state
      )

      if (hasJSX) {
        // add a fancyelement hook to start of render
        path.node.body.body.unshift(
          t.variableDeclaration('const', [
            t.variableDeclarator(
              GLOSS_ID,
              t.identifier('this.glossElement.bind(this)')
            ),
          ])
        )
      }
    },
  }

  const programVisitor = {
    Class(path: Object, state: Object) {
      const node = path.node

      if (!node.decorators || !node.decorators.length) {
        return
      }

      // -- Validate if class is what we're looking for
      //    has some flexibility, looks for any of:
      //       @x  @x()  @x.y  @x.y()

      const decoratorName = (state.opts && state.opts.decoratorName) || 'style'

      const foundDecorator = node.decorators.some(item => {
        if (!item.expression) {
          return false
        }
        // @style
        if (
          item.expression.type === 'Identifier' &&
          item.expression.name === decoratorName
        ) {
          return true
        }
        // @style()
        if (
          item.expression.callee &&
          item.expression.callee.name === decoratorName
        ) {
          return true
        }
        // @style.something()
        if (
          item.expression.callee &&
          t.isMemberExpression(item.expression.callee) &&
          item.expression.callee.object.name === decoratorName
        ) {
          return true
        }
        // @style.something
        if (
          item.expression.object &&
          item.expression.object.name === decoratorName
        ) {
          return true
        }

        return false
      })

      // -- Add a unique var to scope and all of JSX elements
      if (foundDecorator) {
        path.traverse(classBodyVisitor, state)
      }
    },
  }

  return {
    visitor: {
      Program(path: Object, state: Object) {
        path.traverse(programVisitor, state)
      },
    },
  }
}
