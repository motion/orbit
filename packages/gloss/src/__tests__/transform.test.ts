import { transform } from '../transform'
import dedent from 'dedent'

it('works', async () => {
  const { cssText } = await transform(
    dedent`
      import { gloss } from 'gloss'

      export const View = gloss({
        background: [0,0,0],
        transform: {
          y: 10
        }
      })
    `,
    {
      filename: './test.js',
      outputFilename: '../.gloss-cache/test.css',
    },
  )

  expect(cssText).toMatchSnapshot()
})
