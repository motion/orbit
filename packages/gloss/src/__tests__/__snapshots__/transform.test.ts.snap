// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `undefined`;
